rust-runtime/smithy-json/src/deserialize/token.rs

```diff
@@ -0,0 +1,362 @@
+        unescape_string(self.0)
+    }
+}
+
+/// Represents the location of a token
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub struct TokenOffset(pub usize);
+
+impl TokenOffset {
+    /// Creates a custom error from the offset
+    pub fn error(&self, msg: Cow<'static, str>) -> Error {
+        Error::new(ErrorReason::Custom(msg), Some(self.0))
+    }
+}
+
+/// Enum representing the different JSON tokens that can be returned by json_token_iter.
+#[derive(Debug, PartialEq)]
+pub enum Token<'a> {
+    StartArray {
+        offset: TokenOffset,
+    },
+    EndArray {
+        offset: TokenOffset,
+    },
+    ObjectKey {
+        offset: TokenOffset,
+        key: EscapedStr<'a>,
+    },
+    StartObject {
+        offset: TokenOffset,
+    },
+    EndObject {
+        offset: TokenOffset,
+    },
+    ValueBool {
+        offset: TokenOffset,
+        value: bool,
+    },
+    ValueNull {
+        offset: TokenOffset,
+    },
+    ValueNumber {
+        offset: TokenOffset,
+        value: Number,
+    },
+    ValueString {
+        offset: TokenOffset,
+        value: EscapedStr<'a>,
+    },
+}
+
+impl<'a> Token<'a> {
+    pub fn offset(&self) -> TokenOffset {
+        use Token::*;
+        *match self {
+            StartArray { offset } => offset,
+            EndArray { offset } => offset,
+            ObjectKey { offset, .. } => offset,
+            StartObject { offset } => offset,
+            EndObject { offset } => offset,
+            ValueBool { offset, .. } => offset,
+            ValueNull { offset } => offset,
+            ValueNumber { offset, .. } => offset,
+            ValueString { offset, .. } => offset,
+        }
+    }
+
+    /// Builds an error from the token's offset
+    pub fn error(&self, msg: Cow<'static, str>) -> Error {
+        self.offset().error(msg)
+    }
+}
+
+macro_rules! expect_fn {
+    ($name:ident, $typ:ident) => {
+        pub fn $name(token_result: Option<Result<Token<'_>, Error>>) -> Result<(), Error> {
+            match token_result.transpose()? {
+                Some(Token::$typ { .. }) => Ok(()),
+                Some(token) => {
+                    Err(token.error(Cow::Borrowed(concat!("expected ", stringify!($typ)))))
+                }
+                None => Err(Error::new(
+                    ErrorReason::Custom(Cow::Borrowed(concat!("expected ", stringify!($typ)))),
+                    None,
+                )),
+            }
+        }
+    };
+}
+
+expect_fn!(expect_start_object, StartObject);
+expect_fn!(expect_start_array, StartArray);
+
+/// Expects a string or null token, and if its a string, returns its unescaped value.
+pub fn expect_string_or_null(
+    token: Option<Result<Token<'_>, Error>>,
+) -> Result<Option<String>, Error> {
+    match token.transpose()? {
+        Some(Token::ValueNull { .. }) => Ok(None),
+        Some(Token::ValueString { value, .. }) => Ok(Some(value.to_unescaped()?.to_string())),
```

Comment:
I guess you know you're definitely going to need a string at this point so there is no point in keeping at a `Cow`?

Reply:
Exactly. The only time this gets called is when we're moving a string value into a builder.